static/src/javascripts/projects/common/modules/commercial/dfp/dfp-api.js

```diff
@@ -671,37 +622,61 @@ define([
      * ADVERT DOMAIN OBJECTS
      */
 
-    function getAdverts() {
-        return adverts;
+    function getAdvertById(id) {
+        return id in advertIds ? adverts[advertIds[id]] : null;
     }
 
-    function getAdvertArray() {
-        return Object.keys(adverts).map(function (key) {
-            return adverts[key];
+    var waitForAdvert = memoize(function (id) {
+        return new Promise(function (resolve, reject) {
+            var failedAttempts = 5;
+            checkAdvert();
+            function checkAdvert() {
+                var advert = getAdvertById(id);
+                if (!advert) {
+                    failedAttempts -= 1;
+                    if (failedAttempts === 0) {
+                        reject(new Error('Ad ' + id + ' failed to load'));
+                    }
+                    window.setTimeout(checkAdvert, 100);
+                } else {
+                    resolve(advert);
+                }
+            }
         });
+    });
+
+    function trackAdLoad(id) {
+        return waitForAdvert(id).then(function (_) { return _.whenLoaded; });
     }
 
-    function Advert($adSlot) {
-        this.isRendered = false;
-        this.isLoading = false;
-        this.adSlotId = $adSlot.attr('id');
-        this.sizes = getAdBreakpointSizes($adSlot);
-        this.slot = defineSlot($adSlot, this.sizes);
+    function trackAdRender(id) {
+        return waitForAdvert(id).then(function (_) { return _.whenRendered; });
     }
 
-    function getAdBreakpointSizes($adSlot) {
-        var sizes = {};
-        detect.breakpoints.forEach(function (breakpoint) {
-            var data = $adSlot.data(breakpointNameToAttribute(breakpoint.name));
-            if (data) {
-                sizes[breakpoint.name] = createSizeMapping(data);
+    function getAdverts(withHidden) {
+        return Object.keys(advertIds).reduce(function (advertsById, id) {
+            var advert = getAdvertById(id);
+            if (withHidden || !advert.isHidden) {
+                advertsById[id] = advert;
             }
+            return advertsById;
+        }, {});
+    }
+
+    function initAdvert(adSlotNode) {
+        return Object.seal({
```

Comment:
Could this catch people out? new property changes will fail silently. I think it's a good intro to immutablility, but it's really hard to know when I'm dealing with a sealed or non-sealed object.